[JENKINS-34855] AtomicFileWriter should use NIO Atomic renames when available

christ66 opened the pull request:
[JENKINS-34855](https://issues.jenkins-ci.org/browse/JENKINS-34855)

In certain cases the `AtomicWriteFile` is not `Atomic`. In a rare case you can perform a restart in the middle of a move operation, and on disk you will end up with a `0` byte file because the move operation happened in memory, but failed to be written to disk. We need to make sure that these operations are properly synced with disk.

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

*  `Internal:` [JENKINS-34855](https://issues.jenkins-ci.org/browse/JENKINS-34855): `AtomicFileWriter` use _Atomic Moves_ when supported by the underlying filesystem

(Note: this does *not* fix the whole JENKINS-34855 as it has been requested. Should improve the situation a bit, but a followup to try to do a better job with regard to forcing syncing to disk is coming).

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [X] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- ~~[ ] For dependency updates: links to external changelogs and, if possible, full diffs~~

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees esp. @jtnord

christ66 committed:
[JENKINS-34855] Make atomic file write more atomic by using JDK 7 apis.

christ66 committed:
Perform ATOMIC_MOVE when copying the file. If we are unable to perform the ATOMIC_MOVE operation then we fallback to an operation which is supported by all OSes.
Create constructor for charsets.
Delete tempdir instead of destdir.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

christ66 committed:
Finish javadoc deprecated statement.

christ66 committed:
Commit should not perform a delete as it is already performing a move.
Add suffix to create temp file.

christ66 committed:
Add unit test for atomic file writer.

orrc commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,19 +67,29 @@ public AtomicFileWriter(File f) throws IOException {
+         * @param charset
+         *      File charset to write. If null, platform default encoding is chosen.
+         */
+    public AtomicFileWriter(File f, Charset charset) throws IOException {
```
Very minor: crazy Javadoc indentation.

christ66 replied:
@orrc Fixed in f8c9c04

christ66 committed:
Fix javadocs.

christ66 committed:
Do not create the directory if it already exists.

christ66 commented:
Not sure why the first CI build failed (there's a lot failing with the timeout).

jglick commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -56,21 +65,31 @@ public AtomicFileWriter(File f) throws IOException {
-        if (encoding==null)
-            encoding = Charset.defaultCharset().name();
-        core = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(tmpFile),encoding));
+        destFile = f.toPath();
```
:ant: better to move this to the top, followed by

```java
Path dir = destFile.getParent();
```

jglick commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -103,34 +122,45 @@ public void close() throws IOException {
-            }
+        try {
+            // Try to make an atomic move.
+            Files.move(tmpFile, destFile, StandardCopyOption.ATOMIC_MOVE);
```
Do you not _also_ want `REPLACE_EXISTING` here? In case `destFile` already existed? (Is this case tested?)

batmat replied:
>  implementation specific 
So though in practice @jtnord is right, I would agree with @jglick we should use `REPLACE_EXISTING` too.

Because, though in my testing on my machine, it does not make indeed any difference. I think this better conveys the thing we are willing to do. And it seems not putting `REPLACE_EXISTING` could fail on a implementation and **we** would be at fault.

jtnord replied:
it is pointless - it is ignored even if that is what you are attempting to do - it means the person that next reads this code goes erm why are they doing X below and the javadoc says its ignored....
(basically atomic-move is  most likley a posix `rename()` which will overwrite and no other options make any sense.  however this is implementation specific and a provided could pass both down and an OS implementation coudl retrurn an error with "incomatable_flags"

jglick commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -103,34 +122,45 @@ public void close() throws IOException {
+        try {
+            // Try to make an atomic move.
+            Files.move(tmpFile, destFile, StandardCopyOption.ATOMIC_MOVE);
+        } catch (IOException e) {
```
:bug: this should be `AtomicMoveNotSupportedException`.

christ66 replied:
[AtomicMoveNotSupportedException](https://docs.oracle.com/javase/7/docs/api/java/nio/file/AtomicMoveNotSupportedException.html) is a type of [FileSystemException](https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileSystemException.html) which is a type of IOException.

jglick replied:
Yes I know. But if some other kind of `IOException` is thrown, this is not the appropriate response. It should be thrown up. IOW you should be catching the exception type which you actually claim to be handling here.

jtnord replied:
@jglick no. If the file exists a generic IOException can be thrown iff atomic move is suppported and the implementation does not overwrite and it does not implement the *optional* FileAlreadyExistsException

jglick commented on core/src/test/java/hudson/util/AtomicFileWriterTest.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Before
+    public void setUp() throws IOException {
+        af = File.createTempFile("AtomicFileWriter", ".tmp");
```
Use

```java
@Rule
TemporaryFolder tmp = new TemporaryFolder();
```

and you need not bother with `tearDown`.

jglick commented on core/src/test/java/hudson/util/AtomicFileWriterTest.java:
```diff
@@ -0,0 +1,74 @@
+        afw.commit();
+
+        // Then
+        assertTrue(Files.size(af.toPath()) == expectedContent.length());
```
```java
assertEquals(expectedContent.length(), Files.size(af.toPath()));
```

jglick commented on core/src/test/java/hudson/util/AtomicFileWriterTest.java:
```diff
@@ -0,0 +1,74 @@
+
+        // Then
+        assertTrue("File writer did not properly flush to temporary file",
+                Files.size(afw.getTemporaryPath()) == expectedContent.length());
```
use `assertEquals`

jglick reviewed (changes requested):
See some minor issues, but the concept looks good.

christ66 committed:
Move destFile initiation to top.

christ66 committed:
Only catch AtomicMoveNotSupportedException.

christ66 committed:
Use temporary file, and assertEquals in tests.

daniel-beck commented:
@oleg-nenashev Unclear whether you would still consider this WIP

oleg-nenashev commented:
@daniel-beck Nope since there were other changes

oleg-nenashev commented:
But @jglick blocks this PR from what I see.

christ66 commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -103,34 +122,45 @@ public void close() throws IOException {
-            }
+        try {
+            // Try to make an atomic move.
+            Files.move(tmpFile, destFile, StandardCopyOption.ATOMIC_MOVE);
```
This case isn't tested yet. I will write something and change the Exception to handle `FileAlreadyExistsException`.

jtnord replied:
Nope - that would be pointless...

	ATOMIC_MOVE: The move is performed as an atomic file system operation and all other options are ignored

jglick replied:
Where do you see that @jtnord? Not [here](http://docs.oracle.com/javase/8/docs/api/java/nio/file/StandardCopyOption.html#ATOMIC_MOVE).

jtnord replied:
@jglick [here](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#move(java.nio.file.Path,%20java.nio.file.Path,%20java.nio.file.CopyOption...))

jglick reviewed (dismissed):
:bee: (was there another test case coming?)

christ66 commented:
Yes, sorry another test is going to be added tonight.

oleg-nenashev commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -56,21 +65,31 @@ public AtomicFileWriter(File f) throws IOException {
      */
     public AtomicFileWriter(File f, String encoding) throws IOException {
-        File dir = f.getParentFile();
+        this(f, Charset.forName(encoding));
```
🐛 Violates Javadoc. Will be <code>IllegalArgumentException</code> - https://docs.oracle.com/javase/7/docs/api/java/nio/charset/Charset.html#forName(java.lang.String)

batmat replied:
@oleg-nenashev we cannot wrap this in a try-catch because it's a constructor call. Do you mean we simply need to enrich the Javadoc, or do we want to wrap this in some way to avoid throwing this kind of exception.

jtnord replied:
you could move the constructor code to an `init()` style method and wrap the CharsetNotFound with an `IOException`

batmat replied:
@oleg-nenashev @jtnord Addressed: I finally went through a path (no pun intended) that looked better to me. I'm not allowing a null parameter on the new method, and used a ternary on the deprecated shim constructor to defer to the new method.

I'm addressing I think @oleg-nenashev's concern with now a still valid Javadoc: `If null, platform default encoding is chosen.` is still true.
And removed that Javadoc + added `@Nonnull` the new constructor.

oleg-nenashev commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -56,21 +65,31 @@ public AtomicFileWriter(File f) throws IOException {
-            encoding = Charset.defaultCharset().name();
-        core = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(tmpFile),encoding));
+        if (charset==null)
+            charset = Charset.defaultCharset();
```
please use brackets for such cases

batmat replied:
Done

oleg-nenashev commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -103,34 +122,45 @@ public void close() throws IOException {
+    /**
+     * Until the data is committed, this file captures
+     * the written content.
+     *
```
<code>@since TODO</code>

oleg-nenashev commented on core/src/test/java/hudson/util/AtomicFileWriterTest.java:
```diff
@@ -0,0 +1,71 @@
+        // Then
+        assertTrue(Files.notExists(afw.getTemporaryPath()));
+    }
+}
```
line break at the end?

batmat replied:
done

oleg-nenashev reviewed (dismissed):
🐛 for Javadoc violation in the constructor. Also other minor comments

oleg-nenashev commented:
@christ66 gentle ping

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -103,34 +122,45 @@ public void close() throws IOException {
+        try {
+            // Try to make an atomic move.
+            Files.move(tmpFile, destFile, StandardCopyOption.ATOMIC_MOVE);
+        } catch (AtomicMoveNotSupportedException e) {
```
Looks like this was changed and introduced a :bug:

batmat replied:
@jtnord I guess you're saying we need to keep the final fallback 

```java
} catch (IOException x) {
       Files.delete(tmpFile);
       throw x;
}
```

right?

jtnord replied:
no it should catch IOEXception and then try to do the move non-atomically.

then if that fails it should throw the resulting exception (with the previous supressed)

jtnord reviewed (changes requested):
Change the code back to handle ioexception in commit.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

jglick commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -56,21 +65,31 @@ public AtomicFileWriter(File f) throws IOException {
-            dir.mkdirs();
-            tmpFile = File.createTempFile("atomic",null, dir);
+            if (Files.notExists(dir)) {
+                Files.createDirectories(dir);
```
The `if` clause is unnecessary:

> Unlike the `createDirectory` method, an exception is not thrown if the directory could not be created because it already exists.

batmat committed:
Add @since TODO on new method

batmat committed:
Remove unnecessary if

batmat committed:
Wrap if block with brackets

Because brackets should always be there :-)

batmat commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -103,34 +122,45 @@ public void close() throws IOException {
     public void abort() throws IOException {
         close();
-        tmpFile.delete();
+        Files.deleteIfExists(tmpFile);
```
@johnou You mean in case the `close()` has an issue?

johnou replied:
yep

johnou replied:
something like this in case close throws IO exception.

```java
try {
   close();
} finally {
   Files.deleteIfExists(tmpFile);
}
```

johnou replied:
@batmat I think the above approach makes sense in abort and finalize given the javadoc "one way or the other, temporary file should be deleted".

batmat committed:
Force line break

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -56,21 +65,30 @@ public AtomicFileWriter(File f) throws IOException {
+     *
+     * @deprecated Use {@link #AtomicFileWriter(File, Charset)}
      */
     public AtomicFileWriter(File f, String encoding) throws IOException {
```
add `@Deprecated` annotation

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -56,21 +65,30 @@ public AtomicFileWriter(File f) throws IOException {
+        /**
+         * @param charset File charset to write. If null, platform default encoding is chosen.
+         */
+    public AtomicFileWriter(File f, Charset charset) throws IOException {
```
crazy indentiation on the javadoc comment

jtnord replied:
should this actually take a Path to begin with?

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -56,21 +65,30 @@ public AtomicFileWriter(File f) throws IOException {
+         * @param charset File charset to write. If null, platform default encoding is chosen.
+         */
+    public AtomicFileWriter(File f, Charset charset) throws IOException {
+        destFile = f.toPath();
```
f.toPath may throw an [`InvalidPathException`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/InvalidPathException.html) which is not an `IOException`.

batmat replied:
Yeah, but it's a `IllegalArgumentException` which is a runtime anyway. So, well. Can add javadoc, or wrap this all to preserve previous behaviour on previous code paths, and propagate on new ones.

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -103,34 +121,45 @@ public void close() throws IOException {
      * Until the data is committed, this file captures
      * the written content.
+     *
+     * @deprecated Use getTemporaryPath() for JDK 7+
```
this is now for JDK8 as core requires 8 or higher...

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -103,34 +121,45 @@ public void close() throws IOException {
+            // If it falls here that means that Atomic move is not supported by the OS.
+            // In this case we need to fall-back to a copy option which is supported by all OSes.
+            Files.move(tmpFile, destFile, StandardCopyOption.REPLACE_EXISTING);
         }
```
you are missing catching an `IOException` for the case that `StandardCopyOption.ATOMIC_MOVE` **is** supported but that the Atomic Move can not replace an existing file.

Now this should really be a `FileAlreadyExists` exception but this exception is optional so you really do have to work with a generic `IOException` (yay :-o )

batmat committed:
Also annotate with @Deprecated

batmat committed:
Fix excessive indentation

batmat committed:
Jenkins being now Java 8+: no need to warn about 7+

batmat committed:
Rename tmpFile and destFile to *Path now they are Path

batmat committed:
Do not allow null charset on the new constructor

batmat committed:
catch IOException and do a best effort to fall back

And if things go wrong, we try to spit logs to help diagnostics.

batmat committed:
Merge remote-tracking branch 'upstream/master' into JENKINS-34855

# Conflicts:
#	core/src/main/java/hudson/util/AtomicFileWriter.java

batmat committed:
Delete the temp file even if close() throws an exception

batmat commented:
@jtnord I believe I addressed your concern.

@oleg-nenashev @jglick it's ready for another round of review since your last look if possible.

Thanks!

batmat commented:
@johnou I fixed your suggestion, thanks!

batmat commented:
The CI failure is a test failure only on Windows. I'm currently running a Windows build to see if I see that consistently, though as it's apparently a timeout I'm inclined to think it could be unrelated.

oleg-nenashev commented:
I have invalidated my review and requested the new one.
No review ETA commitment

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -108,35 +119,72 @@ public void close() throws IOException {
+        } catch (IOException e) {
+            // If it falls here that can mean many things. Either that the atomic move is not supported,
+            // or something wrong happened. Anyway, let's try to be over-diagnosing
+            LOGGER.log(Level.WARNING, "Unable to move atomically, falling back to non-atomic move.");
```
I had a suspicion that this would cause a lot of logging for people that use certain (CIFS) network shares as Jenkins home
but tested writing to a cifs share and the ATOMIC_MOVE does not report an error.

```
        File source = new File("\\\\nas\\home\\test\\one.txt");
        File target = new File("\\\\nas\\home\\test\\two.txt");
        final Path sourcePath = source.toPath();
        final Path targetPath = target.toPath();
        
        Files.write(sourcePath, "source".getBytes());
        Files.write(targetPath, "target".getBytes());

        assertTrue(source.exists());
        assertTrue(target.exists());

        Files.move(sourcePath, targetPath, StandardCopyOption.ATOMIC_MOVE);

```

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -108,35 +119,72 @@ public void close() throws IOException {
+        try {
+            // Try to make an atomic move.
+            Files.move(tmpPath, destPath, StandardCopyOption.ATOMIC_MOVE);
+        } catch (IOException e) {
```
Possibly would be better to add a specific catch for `AtomicMoveNotSupported` just to have a clearer message (and in this case it is almost expected so we would not need to log it.  Still need to do exactly the same code afterwards though.

batmat replied:
>  Still need to do exactly the same code afterwards though.

yeah, that's why I didn't do that. But can add a private method for that purpose. Though my `if(destPath exists)` I put for that purpose: clarifying what could have gone wrong, but yeah.

jtnord replied:
or just `if (e instanceof AtomicMoveNotSupported` { /* some logging */} else { /* some other logging */}` either works.

batmat replied:
Done

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,25 +61,33 @@ public AtomicFileWriter(File f) throws IOException {
-        File dir = f.getParentFile();
+    @Deprecated
+    public AtomicFileWriter(@Nonnull File f, String encoding) throws IOException {
+        this(f.toPath(), encoding == null ? Charset.defaultCharset() : Charset.forName(encoding));
```
`f.toPath()`  can throw `InvalidPathException` which could cause calling threads to die as it is a newly unexpected exception.

I would put all of this into an init so that you can catch the exception and wrap it in an `IOException` so that you do not potentially cause threads to die due to this change.

batmat replied:
Done. Finally just created a wrapper for the `File.toPath()` method to throw `IOException` instead if an `InvalidPathException` is thrown on this call.

jtnord reviewed (commented):
reservations about `f.toPath()` throwing new runtime exceptions

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -108,35 +119,72 @@ public void close() throws IOException {
+                Files.move(tmpPath, destPath, StandardCopyOption.REPLACE_EXISTING);
+            } catch (IOException e1) {
+                e1.addSuppressed(e);
+                LOGGER.log(Level.SEVERE, "Unable to move {0} to {1}. Attempting to delete {1} and abandoning.",
```
:bug:  ` Attempting to delete {1}` should be ` Attempting to delete {0}`

batmat replied:
Fixed

batmat committed:
Fix wrong log parameter index

batmat committed:
Wraps InvalidPathException in IOException if thrown

To keep backward compatibility and avoid potentially killing
threads that would be using this code.

batmat committed:
Moar diagnostics in case atomic move fails

batmat commented:
@jtnord "reservations about f.toPath() throwing new runtime exceptions" addressed.

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,27 +64,51 @@ public AtomicFileWriter(File f) throws IOException {
-            encoding = Charset.defaultCharset().name();
+    @Deprecated
+    public AtomicFileWriter(@Nonnull File f, @Nullable String encoding) throws IOException {
+        this(toPath(f), encoding == null ? Charset.defaultCharset() : Charset.forName(encoding));
```
Not sure why I only noticed this now but 
`Charset.forName()` throws
* `IllegalCharsetNameException` - If the given charset name is illegal
* `IllegalArgumentException` - If the given charsetName is null
* `UnsupportedCharsetException` - If no support for the named charset is available in this instance of the Java virtual machine

batmat replied:
Yes, should have added a comment here myself preempively: the previous code path was going through `core = new BufferedWriter(new OutputStreamWriter(Files.newOutputStream(tmpFile.toPath()), encoding));`

And `new OutputStreamWriter(Files.newOutputStream(tmpFile.toPath()), encoding)` ends up calling Charset.forName() already.

```java
public OutputStreamWriter(OutputStream out, String charsetName)
        throws UnsupportedEncodingException
    {
        super(out);
        if (charsetName == null)
            throw new NullPointerException("charsetName");
        se = StreamEncoder.forOutputStreamWriter(out, this, charsetName);
    }
```

which calls:

```java
public static StreamEncoder forOutputStreamWriter(OutputStream var0, Object var1, String var2) throws UnsupportedEncodingException {
        String var3 = var2;
        if (var2 == null) {
            var3 = Charset.defaultCharset().name();
        }

        try {
            if (Charset.isSupported(var3)) {
                return new StreamEncoder(var0, var1, Charset.forName(var3));
            }
        } catch (IllegalCharsetNameException var5) {
            ;
        }

        throw new UnsupportedEncodingException(var3);
    }
```

Granted, **if** the caller put a non existing encoding, they would previously get an `UnsupportedEncodingException` because then `Charset.isSupported(var3)` may have returned false.

But, I don't imagine anyone catching `UnsupportedEncodingException` to handle it instead of just using an existing encoding, and hence would be hit by such change?

jtnord replied:
OK!

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,27 +64,51 @@ public AtomicFileWriter(File f) throws IOException {
+     * @return the path for that file
+     * @see File#toPath()
+     */
+    private static Path toPath(@Nonnull File file) throws IOException {
```
NIT static functions normally come after constructors in the style guide.

oleg-nenashev replied:
+1 I would also put this method to a utility class. we have lots of other cases where it is required

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,27 +64,51 @@ public AtomicFileWriter(File f) throws IOException {
     }
```
@batmat as discussed where you going to deprecate this?

batmat replied:
@jtnord not in this PR no. It is out of scope to me, so I can possibly file it separately, but I don't want to _risk_ triggering another serie of comments and fixes to do on something that is not central to this very change.

jglick commented:
[Example test failure](https://ci.jenkins.io/job/Core/job/jenkins/job/PR-2548/24/testReport/hudson.util/SecretRewriterTest/Linux___Linux_Publishing___recursionDetection/):

```
java.io.IOException: Failed to create a temporary file in …/core/target/junit6092863236215899237/t/a/symlink
	at hudson.util.AtomicFileWriter.<init>(AtomicFileWriter.java:106)
	at hudson.util.AtomicFileWriter.<init>(AtomicFileWriter.java:73)
	at hudson.util.SecretRewriter.rewrite(SecretRewriter.java:85)
	at …
```

johnou commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,27 +64,51 @@ public AtomicFileWriter(File f) throws IOException {
+        Path dir = this.destPath.getParent();
+        try {
+            Files.createDirectories(dir);
+            tmpPath = Files.createTempFile(dir, "atomic", "tmp");
```
@batmat what's the lifecycle of this? Be weary that certain Redhat distributions periodically destroy files in the tmp directory [1].

[1]

The file /etc/cron.daily/tmpwatch defines the way tmpwatch is called.

```
#! /bin/sh
flags=-umc
/usr/sbin/tmpwatch "$flags" -x /tmp/.X11-unix -x /tmp/.XIM-unix \
    -x /tmp/.font-unix -x /tmp/.ICE-unix -x /tmp/.Test-unix \
    -X '/tmp/hsperfdata_*' 10d /tmp
```

johnou replied:
If it is short lived (for the operation) it is fine, just thought I would share the above experience :)

jtnord replied:
This is not in the temp directory but the same directory as the target file (generally somewhere under JENKINS_HOME)

johnou replied:
Even better!

batmat commented:
FWIW, ran `mvn -V -B -e install  -DfailIfNoTests=false -Dtest=RunRangeCommandTest,WaitNodeOfflineCommandTest,SecretRewriterTest` on a private Windows setup some times, and never had a failure like here.

I had `hudson.UtilTest.testDeleteRecursive_onWindows` failing only once among many runs, and I suspect this is not even related and _just_ a IO-performance related issue (i.e. ~flaky-friendly).

IOW, fun journey with "random" failures. Looking as this code is very central and we need to really assess if this could be visibly impacting something or not. (BTW, possibly it would, but it could be OK as we are here trying to go for more integrity, before performances...)

batmat committed:
Add an if to avoid crashing on symlinks to dirs

Files.createDirectories(dir) is said to not fail if directory already
exists.
*But* this will still fail if `dir` is actually a symlink to a
directory...
This is documented in the Javadoc indeed, even if probably not really
one would expect...

See https://bugs.openjdk.java.net/browse/JDK-8130464

Without this fix:
```
[INFO] Running hudson.util.SecretRewriterTest
Cycle detected: /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/c/symlink
Scanning /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/c/foo.xml
Rewritten /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/c/foo.xml
Rewritten /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/c/d/foo.xml
Rewritten /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/c/d/e/foo.xml
ERROR: Failed to rewrite /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/a/symlink/foo.xml
java.nio.file.FileAlreadyExistsException: /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/a/symlink
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:88)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384)
        at java.nio.file.Files.createDirectory(Files.java:674)
        at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781)
        at java.nio.file.Files.createDirectories(Files.java:727)
        at hudson.util.AtomicFileWriter.<init>(AtomicFileWriter.java:103)
Caused: java.io.IOException: Failed to create a temporary file in /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/a/symlink
        at hudson.util.AtomicFileWriter.<init>(AtomicFileWriter.java:106)
        at hudson.util.AtomicFileWriter.<init>(AtomicFileWriter.java:73)
        at hudson.util.SecretRewriter.rewrite(SecretRewriter.java:85)
        at hudson.util.SecretRewriter.rewriteRecursive(SecretRewriter.java:166)
        at hudson.util.SecretRewriter.rewriteRecursive(SecretRewriter.java:176)
        at hudson.util.SecretRewriter.rewriteRecursive(SecretRewriter.java:176)
        at hudson.util.SecretRewriter.rewriteRecursive(SecretRewriter.java:141)
        at hudson.util.SecretRewriter$rewriteRecursive$0.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
        at hudson.util.SecretRewriterTest.recursionDetection(SecretRewriterTest.groovy:110)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:236)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:386)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:323)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:143)
Rewritten /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/a/foo.xml
Cycle detected: /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/b/symlink
Rewritten /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t/b/foo.xml
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.097 s <<< FAILURE! - in hudson.util.SecretRewriterTest
[ERROR] recursionDetection(hudson.util.SecretRewriterTest)  Time elapsed: 0.176 s  <<< FAILURE!
Assertion failed:

assert 6==sw.rewriteRecursive(t, st)
        | |  |                |  |
        | |  5                |  hudson.util.StreamTaskListener@34a2d6e0
        | |                   /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t
        | hudson.util.SecretRewriter@2525a5b8
        false

        at org.codehaus.groovy.runtime.InvokerHelper.assertFailed(InvokerHelper.java:404)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.assertFailed(ScriptBytecodeAdapter.java:650)
        at hudson.util.SecretRewriterTest.recursionDetection(SecretRewriterTest.groovy:110)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:236)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:386)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:323)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:143)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   SecretRewriterTest.recursionDetection:110 assert 6==sw.rewriteRecursive(t, st)
        | |  |                |  |
        | |  5                |  hudson.util.StreamTaskListener@34a2d6e0
        | |                   /home/tiste/dev/JENKINS/jenkins/core/target/junit7537647214741745549/t
        | hudson.util.SecretRewriter@2525a5b8
        false
[INFO]
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0
```

batmat commented:
So after a bisection, I finally found the reason of the failure, pushed in https://github.com/jenkinsci/jenkins/pull/2548/commits/1396c7f774e765f743ab0ffce7bc39bfac76d357

Basically, `Files.createDirectories(dir)` must still be protected with a condition because it will fail on symlinks to directories...

batmat committed:
Retry assertion for 10 seconds before failing

Also reduce the time 250 ms (instead of 1 second previously) between
retries to hopefully reduce the test duration on quick enough setups,
and still degrade nicely on less performing ones.

johnou commented on test/pom.xml:
```diff
@@ -198,6 +198,11 @@ THE SOFTWARE.
+    <dependency>
+      <groupId>org.awaitility</groupId>
+      <artifactId>awaitility</artifactId>
+      <version>3.0.0</version>
```
missing test scope.

batmat replied:
Thanks a lot, fixed.

batmat committed:
Fix RequireUpperBoundDeps issue

```
WARNING] Rule 3: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps fai
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for org.objenesis:objenesis:2.1 paths
+-org.jenkins-ci.main:test:2.92-SNAPSHOT
  +-org.mockito:mockito-core:1.10.19
    +-org.objenesis:objenesis:2.1
and
+-org.jenkins-ci.main:test:2.92-SNAPSHOT
  +-org.awaitility:awaitility:3.0.0
    +-org.objenesis:objenesis:2.5.1
```

batmat committed:
Try to get the future for 10 seconds before timing out

Should make the test more robust on variously performing setups.

batmat committed:
Log exception in case of failure

I see we enter that catch, and the `else` in a failing test.
But I'm somewhat stuck since I miss both the stack, and even at
least the exception being thrown and not being an
`AtomicMoveNotSupportedException`.

It's possibly an `IOException` or some subclass. But still having the
stack could help understand why the test fails, and provide more
data for diagnosis in the future in production use.

```
=== Starting hudson.cli.RunRangeCommandTest
   0.007 [id=96]	WARNING	o.jvnet.hudson.test.JenkinsRule#before: Jenkins.theInstance was not cleared by a previous test, doing that now
   0.018 [id=96]	INFO	o.jvnet.hudson.test.JenkinsRule#createWebServer: Running on http://localhost:45833/jenkins/
   0.037 [id=15]	WARNING	jenkins.model.Jenkins#cleanUp: This instance is no longer the singleton, ignoring cleanUp()
   0.345 [id=96]	WARNING	hudson.util.AtomicFileWriter#commit: Unable to move atomically, falling back to non-atomic move.
   0.345 [id=96]	SEVERE	hudson.util.AtomicFileWriter#commit: Unable to move /home/jenkins/workspace/Core_jenkins_PR-2548-5VRJ4BBBSEGVOZEA7OVAL2YLTLCBL55RB4LDQIZN56WJZD4M6ZKA/test/target/jenkinsTests.tmp/jenkins333129509690162870test/atomic9184049448662376708tmp to /home/jenkins/workspace/Core_jenkins_PR-2548-5VRJ4BBBSEGVOZEA7OVAL2YLTLCBL55RB4LDQIZN56WJZD4M6ZKA/test/target/jenkinsTests.tmp/jenkins333129509690162870test/secret.key. Attempting to delete /home/jenkins/workspace/Core_jenkins_PR-2548-5VRJ4BBBSEGVOZEA7OVAL2YLTLCBL55RB4LDQIZN56WJZD4M6ZKA/test/target/jenkinsTests.tmp/jenkins333129509690162870test/atomic9184049448662376708tmp and abandoning.
```

batmat committed:
Do not force SYNC

It might be preferrable from a consistency standpoint,
but many CLI tests start failing when you force this.

So, as the previous behaviour was not forcing the sync, this seems
to show there could be unintended behavioural changes in effect.

Possibly, we'll want to address this later, but as the main goal
here was to make commit() atomic as much as possible, fixing
this is probably another story anyway.

Note: I was *never* able to reproduce those failures on my machine in
a normal env dev. It was only visible in CI.
I finally managed to make it reproducible (but still a bit randomly) by
using Docker resource constraints, using something like this:

* cd to jenkins local clone
* `docker run -ti -v m2repo:/root/.m2/repository -v $PWD:/work -v ~/.m2/settings.xml:/root/.m2/settings.xml:ro -v ~/.m2/settings-security.xml:/root/.m2/settings-security.xml:ro --device-write-iops /dev/mapper/fedora-home:40 --device-read-iops /dev/mapper/fedora-home:40  --device-write-bps /dev/mapper/fedora-home:10m --device-read-bps /dev/mapper/fedora-home:10m maven:3.5.2-jdk-8 bash`
* cd /work
* `mvn clean install -Dtest=WaitNodeOfflineCommandTest,RunRangeCommandTest -Dfindbugs.skip=true  -DfailIfNoTests=false -Dskip.npm=true`

Failure example:

```
ERROR] Tests run: 6, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 55.68OfflineCommandTest
[ERROR] waitNodeOfflineShouldSucceedOnDisconnectingNode(hudson.cli.WaitNodeOf  <<< FAILURE!
java.lang.AssertionError:

Expected: <true>
     but: was <false>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
        at hudson.cli.WaitNodeOfflineCommandTest.waitNodeOfflineShouldSucceedndTest.java:128)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorIm
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAc
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(Framew
        at org.junit.internal.runners.model.ReflectiveCallable.run(Reflective
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(Framewor
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(Invoke
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefor
        at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:542)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatem
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatem
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.lang.Thread.run(Thread.java:748)

[ERROR] waitNodeOfflineShouldSucceedOnDisconnectedNode(hudson.cli.WaitNodeOff <<< FAILURE!
java.lang.AssertionError
```

batmat commented:
Finally figured out the cause of the failure, fixed in the small https://github.com/jenkinsci/jenkins/pull/2548/commits/465f85e34d23addf1a7f90bcd28e39ae14e509ff

It's ready for last review, hopefully last round before it's ready to merge, now the build succeeds.

amuniz commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -108,35 +139,77 @@ public void close() throws IOException {
+            }
+
+            if (destPath.toFile().exists()) {
+                LOGGER.log(Level.WARNING, "The target file {0} was already existing?!?", destPath);
```
How is this an exceptional case? I think the most common use case is `destPath.toFile().exists() == true`, isn't it?

batmat replied:
@amuniz no, it should not be common actually. See https://github.com/jenkinsci/jenkins/pull/2548#discussion_r84765531

If the underlying FS supports atomic move, it will anyway also replace the file, so we wouldn't even enter that catch.

Now I retrospectively agree the "?!?" is probably a bit too much, though I still want to keep that log overall for more diagnosability.

jtnord replied:
NIT: [`Files.exists(destPath)`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#exists-java.nio.file.Path-java.nio.file.LinkOption...-)

batmat replied:
There seems to be a bug in the JDK about the performances of that method, fixed only in JDK9, that's why I used the old variant.
https://bugs.openjdk.java.net/browse/JDK-8154077
https://bugs.openjdk.java.net/browse/JDK-8153414

It was reported to me by SonarLint I use in my IDE: https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3725

batmat commented:
@johnou it looks like you're still in "requesting change" state. Care to re-review and possibly approve if this now suits you? Thanks

jtnord reviewed (changes requested):
without the `SYNC` flag this is not Atomic.  which is the whole point of this change.  Otherwise you are just changing the API to use `java.nio` which just adds better diagnostics upon failure.

ie. the lack of the `SYNC` flag is what causes us to have zero byte files if jenkins crashes causing data loss because the file metadata is written but the content is not.

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -108,7 +107,7 @@ public AtomicFileWriter(@Nonnull Path destinationPath, @Nonnull Charset charset)
         }
 
-        core = Files.newBufferedWriter(tmpPath, charset, StandardOpenOption.SYNC);
+        core = Files.newBufferedWriter(tmpPath, charset);
```
This change means that the referenced JIRA is not fixed.

johnou replied:
@batmat I'm happy to take another look once SYNC is restored, was this the only failing unit test? https://github.com/jenkinsci/jenkins/pull/2548/commits

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,27 +63,53 @@ public AtomicFileWriter(File f) throws IOException {
+        this.destPath = destinationPath;
+        Path dir = this.destPath.getParent();
+        try {
+            if (Files.notExists(dir)) { // This is required for dir symlink handling, see JDK-8130464
```
should this be more defensive and check that if the file does exist that it is either a symlink to a directory or a directory?

```java
if (Files.exists(dir) && ((!isDirectory(dir)))  {
 throw new IOException(dir + " exists and is neither a direcotry nor a symlink to a directory");
}
else {
  Files.createDirectories(dir);
}
```

oleg-nenashev replied:
+1

oleg-nenashev commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,27 +63,53 @@ public AtomicFileWriter(File f) throws IOException {
+     */
+    public AtomicFileWriter(@Nonnull Path destinationPath, @Nonnull Charset charset) throws IOException {
+        if (charset == null) { // be extra-defensive if people don't care
+            throw new IllegalArgumentException("charset is null");
```
Javadoc will start reporting warnings here, can be polished later

oleg-nenashev commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -108,35 +139,77 @@ public void close() throws IOException {
+                Files.move(tmpPath, destPath, StandardCopyOption.REPLACE_EXISTING);
+            } catch (IOException e1) {
+                e1.addSuppressed(e);
+                LOGGER.log(Level.SEVERE, "Unable to move {0} to {1}. Attempting to delete {0} and abandoning.",
```
are these guys really severe? I doubt. Maybe just WARNING?

batmat replied:
fixed.

oleg-nenashev commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -108,35 +139,77 @@ public void close() throws IOException {
+                    Files.deleteIfExists(tmpPath);
+                } catch (IOException e2) {
+                    e2.addSuppressed(e1);
+                    LOGGER.log(Level.SEVERE, "Unable to delete {0}, good bye then!", tmpPath);
```
same

oleg-nenashev commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -46,9 +49,11 @@
     private final Writer core;
-    private final File tmpFile;
-    private final File destFile;
+    private final Path tmpPath;
```
Nonnull from what I see

oleg-nenashev reviewed (approved):
The pull requests needs some polishing IMHO, but generally I do not see any major issues here. 🐝

batmat committed:
Poor man regression testing

batmat committed:
Reduce Log levels

Not that critical, but still can keep them as we should
generally be on FS that support atomic moves. Hence not
enter that catch.

If proved otherwise, we'll reconsider adjusting this verbosity.

batmat committed:
Be more defensive if provided destPath is wrong

The path could already exist, *but* not be a directory.
Nice catch James.

Associated tests.

batmat commented:
@jtnord addressed partly your feedback. 

As discussed, using `SYNC` is impossible as-is for performance impact reasons. Because then, the time needed to do some operations, and the time needed for some tests just skyrockets.

Example:

```
With Sync

-------------------------------------------------------------------------------
Test set: hudson.cli.RunRangeCommandTest
-------------------------------------------------------------------------------
Tests run: 36, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 198.87 s - in hudson.cli.RunRangeCommandTest


WithOUT SYNC

-------------------------------------------------------------------------------
Test set: hudson.cli.RunRangeCommandTest
-------------------------------------------------------------------------------
Tests run: 36, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 23.615 s - in hudson.cli.RunRangeCommandTest
```

I also pushed https://github.com/jenkinsci/jenkins/pull/2548/commits/927192fc00de4b9cbfed9d1008137af2d9f509aa to watch out possible huge regressions like this in the future. It took me too much time to desintricate the actual cause for those CLI related tests failures. Hopefully this test, if ever failing, will more quickly and obviously point future PR submitters to such issue.

Now, I think there's already some value provided in this PR, like using `NIO` new API for atomic moves.

So, though I agree the disk caching/sync issue is still there, I think we could consider merging this PR to move forward.

Then I will file a followup to probably use `FileChannel.close()` to improve/fix the current disk sync issue. 
IMO this has some pros:
* will allow the current change to be released and possibly provide some feedback (no bad feedback would be already some feedback :-))
* will make reviewing the followup way easier for everywhere, being smaller and more focused on the missing part.

WDYT?

batmat commented on test/src/test/java/hudson/util/AtomicFileWriterPerfTest.java:
```diff
@@ -0,0 +1,34 @@
+     * <li>On a modern SSD in 2017, it takes less than a second to run.</li>
+     * <li>Using Docker resource constraints, and setting in read&write to IOPS=40 and BPS=10m (i.e. roughly worse than
+     * an old 5400 RPM hard disk), this test takes 25 seconds</li>
+     * </ul>
```
FTR, takes roughly 1 second too in ci.jenkins.io:

![image](https://user-images.githubusercontent.com/223853/33368503-31b5f1ee-d4f2-11e7-879a-e771d987ba8a.png)

jtnord commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,27 +63,62 @@ public AtomicFileWriter(File f) throws IOException {
+            if (Files.isSymbolicLink(dir)) {
+                LOGGER.log(Level.CONFIG, "{0} is a symlink to a directory", dir);
+            } else {
+                Files.createDirectories(dir); // Cannot be called on symlink, so we are pretty defensive...
```
ref: https://bugs.openjdk.java.net/browse/JDK-8130464

jtnord reviewed (approved):
improves the window for not being atomic, but can still leave zero byte files if the OS crashes.

OK to address in a follow-up

batmat commented:
@reviewbybees done

batmat commented:
I will leave it open until the next day or so, and will merge if nobody objects in the meantime. 

Working on the followup.

Thanks everyone for the review.

amuniz commented on core/src/test/java/hudson/util/AtomicFileWriterTest.java:
```diff
@@ -0,0 +1,108 @@
+
+        final Path childFileInSymlinkToDir = Paths.get(zeSymlink.toString(), "childFileInSymlinkToDir");
+
+        new AtomicFileWriter(childFileInSymlinkToDir, Charset.forName("UTF-8"));
```
No checks after this? So this is testing no exception is thrown but nothing about `core` and `tmpPath` correctly created.

batmat replied:
Yeah, it was really added exclusively in the last commit with the added `if` to check we don't fail on symlinks (which was actually covered by the `SecretRewriterTest`, but taking longer to understand when it fails).

Other tests are expected to cover deeper usages.

amuniz commented on core/src/test/java/hudson/util/AtomicFileWriterTest.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    @Test
+    public void commitToFile() throws Exception {
```
What about trying to reproduce the non-Atomic write issue in a test?

I think you could try to generate an `IndexOutOfBoundsException` using `write(char cbuf[], int off, int len)`, then catch it and check the original file is intact.

amuniz reviewed (commented):
Looks good. Just a suggestion of an additional test.

batmat commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -108,35 +148,77 @@ public void close() throws IOException {
+            }
+
+            if (destPath.toFile().exists()) {
+                LOGGER.log(Level.INFO, "The target file {0} was already existing", destPath);
```
I tend to agree, but my understanding and testing seems to tell me we should rarely enter that block. Most FS would support atomic moves AIUI. 
One known case, even when the default FS supports it, is when moving file from say one partition to another. But due to how `AtomicFileWrite` creates the temp file along the destination one, it's not something that would happen here.

batmat commented:
Going to merge most probably in the next hour or so as I'm about to file followups (so diff will be huge/useless if I don't merge before) and the code got approved already. It didn't change much since two days (mostly only added tests) after Oleg's review, and James and John approved too in the meantime.

batmat commented:
And thanks a lot @christ66 !

batmat commented:
The https://github.com/jenkinsci/jenkins/pull/3170 followup is ready for review.

dwnusbaum commented on core/src/main/java/hudson/util/AtomicFileWriter.java:
```diff
@@ -58,27 +63,62 @@ public AtomicFileWriter(File f) throws IOException {
+        }
+
+        try {
+            tmpPath = Files.createTempFile(dir, "atomic", "tmp");
```
@batmat I think this is the cause of [JENKINS-48407](https://issues.jenkins-ci.org/browse/JENKINS-48407). In short, nio `Files#createTempFile` sets `0600` permissions on the new file, while the old io `File#createTempFile` sets `0666` (With default `umask` this becomes `0644`), so all files written through AFW now have different permissions. I came up with a quick patch in the comments of the linked issue, could you take a look?
